// web/backend/build/routers/productRouter.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.productRouter = void 0;
exports.getRelatedProducts = getRelatedProducts;
const express_1 = __importDefault(require("express"));
const express_async_handler_1 = __importDefault(require("express-async-handler"));
const productModel_1 = require("../models/productModel");
const utils_1 = require("../utils");
exports.productRouter = express_1.default.Router();
// --- CONTROLLER cho Sáº¢N PHáº¨M LIÃŠN QUAN ---
async function getRelatedProducts(req, res) {
    const { category, exclude, limit } = req.query;
    if (!category) {
        res.status(400).send({ message: 'Category parameter is required' });
        return;
    }
    const limitNum = parseInt(limit) || 4;
    try {
        const products = await productModel_1.ProductModel.find({
            category: category,
            _id: { $ne: exclude },
            countInStock: { $gt: 0 }
        })
            .limit(limitNum)
            .sort({ rating: -1, createdAt: -1 });
        res.send(products);
    }
    catch (error) {
        console.error("Error fetching related products:", error);
        res.status(500).send({ message: 'Failed to fetch related products.' });
    }
}
// âœ… GET /api/products â†’ public with filters & sorting
// ðŸ’¡ Äáº·t route generic TRÆ¯á»šC routes specific
exports.productRouter.get('/', (0, express_async_handler_1.default)(async (req, res) => {
    try {
        const { category, minPrice, maxPrice, rating, sortBy, inStock } = req.query;
        const filter = {};
        if (category) {
            if (category === 'Phone') {
                filter.$or = [
                    { category: 'Iphone' },
                    { category: 'Samsung' },
                    { category: 'Xiaomi' },
                    { category: 'Honor' },
                ];
            }
            else {
                filter.$or = [
                    { category: category },
                    { brand: category }
                ];
            }
        }
        if (minPrice || maxPrice) {
            filter.price = {};
            if (minPrice)
                filter.price.$gte = Number(minPrice);
            if (maxPrice)
                filter.price.$lte = Number(maxPrice);
        }
        if (rating) {
            filter.rating = { $gte: Number(rating) };
        }
        if (inStock === 'true') {
            filter.countInStock = { $gt: 0 };
        }
        let sort = { createdAt: -1 };
        if (sortBy === 'price-low')
            sort = { price: 1 };
        else if (sortBy === 'price-high')
            sort = { price: -1 };
        else if (sortBy === 'rating')
            sort = { rating: -1 };
        const products = await productModel_1.ProductModel.find(filter).sort(sort);
        res.json(products);
    }
    catch (error) {
        console.error('Product API Error:', error);
        res.status(500).json({
            message: 'Error fetching products',
            error: error instanceof Error ? error.message : String(error)
        });
    }
}));
// âœ… GET /api/products/slug/:slug â†’ public
exports.productRouter.get('/slug/:slug', (0, express_async_handler_1.default)(async (req, res) => {
    const product = await productModel_1.ProductModel.findOne({ slug: req.params.slug });
    if (product) {
        res.json(product);
    }
    else {
        res.status(404).json({ message: 'Product Not Found' });
    }
}));
// âœ… GET /api/products/related â†’ public
exports.productRouter.get('/related', (0, express_async_handler_1.default)(getRelatedProducts));
// âœ… GET /api/products/by-rating â†’ public
exports.productRouter.get('/by-rating', (0, express_async_handler_1.default)(async (req, res) => {
    const { limit = 10 } = req.query;
    const products = await productModel_1.ProductModel.find()
        .sort({ rating: -1, numReviews: -1 })
        .limit(Number(limit));
    res.json(products);
}));
// âœ… GET /api/products/admin â†’ admin only (THÃŠM ROUTE NÃ€Y)
exports.productRouter.get('/admin', utils_1.isAuth, utils_1.isAdmin, (0, express_async_handler_1.default)(async (req, res) => {
    console.log('Fetching all products for admin');
    const products = await productModel_1.ProductModel.find();
    console.log('Found products:', products.length);
    res.json(products);
}));
// âœ… POST /api/products â†’ create product (admin only)
exports.productRouter.post('/', utils_1.isAuth, utils_1.isAdmin, (0, express_async_handler_1.default)(async (req, res) => {
    const { name, brand, category, description, price, countInStock, image, rating, numReviews, variants } = req.body;
    // âœ… Validation
    if (!name || !brand || !category || !image) {
        res.status(400).json({
            message: 'Missing required fields: name, brand, category, image'
        });
        return;
    }
    const product = new productModel_1.ProductModel({
        name,
        slug: name.toLowerCase().replace(/\s+/g, '-') + '-' + Date.now(),
        brand,
        category,
        description,
        price: Number(price) || 0,
        countInStock: Number(countInStock) || 0,
        image,
        rating: Number(rating) || 0,
        numReviews: Number(numReviews) || 0,
        variants: variants || [],
    });
    console.log('Creating product:', product);
    const created = await product.save();
    console.log('Product created:', created);
    res.status(201).json(created);
}));
// âœ… GET /api/products/:id â†’ get single product
exports.productRouter.get('/:id', (0, express_async_handler_1.default)(async (req, res) => {
    console.log('Fetching product by ID:', req.params.id);
    const product = await productModel_1.ProductModel.findById(req.params.id);
    if (product) {
        res.json(product);
    }
    else {
        res.status(404).json({ message: 'Product not found' });
    }
}));
// âœ… PUT /api/products/:id â†’ update product (admin only)
exports.productRouter.put('/:id', utils_1.isAuth, utils_1.isAdmin, (0, express_async_handler_1.default)(async (req, res) => {
    console.log('Updating product:', req.params.id);
    const product = await productModel_1.ProductModel.findById(req.params.id);
    if (product) {
        product.name = req.body.name || product.name;
        product.brand = req.body.brand || product.brand;
        product.category = req.body.category || product.category;
        product.description = req.body.description || product.description;
        product.price = req.body.price !== undefined ? Number(req.body.price) : product.price;
        product.countInStock = req.body.countInStock !== undefined ? Number(req.body.countInStock) : product.countInStock;
        product.image = req.body.image || product.image;
        product.rating = req.body.rating !== undefined ? Number(req.body.rating) : product.rating;
        product.numReviews = req.body.numReviews !== undefined ? Number(req.body.numReviews) : product.numReviews;
        if (req.body.variants && Array.isArray(req.body.variants)) {
            product.variants = req.body.variants;
        }
        const updated = await product.save();
        console.log('Product updated:', updated);
        res.json(updated);
    }
    else {
        res.status(404).json({ message: 'Product not found' });
    }
}));
// âœ… DELETE /api/products/:id â†’ delete product (admin only)
exports.productRouter.delete('/:id', utils_1.isAuth, utils_1.isAdmin, (0, express_async_handler_1.default)(async (req, res) => {
    console.log('Deleting product:', req.params.id);
    const product = await productModel_1.ProductModel.findByIdAndDelete(req.params.id);
    if (product) {
        res.json({ message: 'Product deleted', product });
    }
    else {
        res.status(404).json({ message: 'Product not found' });
    }
}));
